### ISSUE TITLE
Can't deploy all functions consistently. Fails most of the time.

### ISSUE THREAD

**Issue by MarwanZeitoun** [firebase/firebase-tools#3919]
Hello; 

I have a project with 169 cloud functions and need to set up automatic deployment to firebase in GitLab, but the deployment is always failed. 

I tried with various node versions, also with various access tokens, but always got the same error.
Locally one of the developers is able to deploy without any issue. But locally in my environment, it fails. Even when replicating the versions of node and firebase.

here is my .yml file
`
image: node:14.16.1
before_script:
  - cd functions
  - npm I
  
 script:
    - npx firebase use way-backend
    - npx firebase deploy --only functions --token $FIREBASE_TOKEN
  only:
    refs:
      - master
    changes:
      - functions/**/*
` 

first warnings like this  with many functions
`
⚠ functions: got "Quota Exceeded" error while trying to update projects/xproject/locations/us-central1/functions/xxxxx. Waiting to retry...
`
and then  got this one
`
Functions deploy had errors with the following functions:
	function1(us-central1)
	function2(us-central1)
	.... etc
Because there were errors creating or updating functions, the following functions were not deleted
To delete these, use firebase functions:delete
`
and not all functions throw the error only a few functions
and finally got this one

`
i  functions: cleaning up build files...
Cleaning up project directory and file based variables
00:01
ERROR: Job failed: exit code 1
`

Thanks in advance

**Comment by abeisgoat** [comment]
Hey we're aware of issues when deploying over 100 functions, it's best to deploy the functions you changed specifically instead of the whole lot when you're at this scale.

**Comment by MarwanZeitoun** [comment]
> Hey we're aware of issues when deploying over 100 functions, it's best to deploy the functions you changed specifically instead of the whole lot when you're at this scale.

usually, we deploy only changed functions, but we need to find a way to set up CI/CD on this project, also when we deploy partially the config changes are still undeployed!

**Comment by ggirotto** [comment]
@abeisgoat I'm also facing a quota issue, although I'm trying to deploy just 20 functions. Can you confirm if it's related or it's another issue? The output that I have in Functions dashboard is:

```
{"@type":"type.googleapis.com/google.cloud.audit.AuditLog","status":{"code":8,"message":"Build failed: Quota exceeded for quota metric 'Build Create requests' and limit 'Build Create requests per minute' of service 'cloudbuild.googleapis.com' for consumer 'project_number:<NUMBER>'."},"authenticationInfo":{"principalEmail":"<EMAIL>"},"serviceName":"cloudfunctions.googleapis.com","methodName":"google.cloud.functions.v1.CloudFunctionsService.UpdateFunction","resourceName":"projects/ticker-dev-4d59e/locations/southamerica-east1/functions/<FUNCTION_NAME>"} 
```

**Comment by yendiscart** [comment]
I have the same issue.


```
Error: Functions did not deploy properly.

Having trouble? Try firebase [command] --help
mac@bogon functions % firebase deploy --only "functions:sendMessageNotification,functions:unsubscribeUserFromBroadcast,functions:setStatusCount,functions:resubscribeUserToBroadcasts,functions:addUserToGroup,functions:deviceIdChanged,functions:sendMessageToBroadcast,functions:participantRemoved,functions:groupInfoChanged,functions:participantAdded,functions:getVirgilJwt,functions:sendUnDeliveredNotifications,functions:unsubscribeUserFromTopicOnDelete,functions:saveUidOnLogin,functions:sendNewCallNotification,functions:deleteMessageForGroup,functions:indexNewCall,functions:groupAdminChanged,functions:deleteMessageForBroadcast,functions:indexNewGroupCall,functions:groupEvents,functions:getTime,functions:sendMessagesForGroups,functions:deleteMessage,functions:indexPKToken,functions:subscribeToBroadcast" 

=== Deploying to 'ichat-daac0'...

i  deploying functions
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
i  functions: preparing functions directory for uploading...
i  functions: packaged functions (73.82 KB) for uploading
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 10 function sendMessageNotification(us-central1)...
i  functions: updating Node.js 10 function sendMessagesForGroups(us-central1)...
i  functions: updating Node.js 10 function participantAdded(us-central1)...
i  functions: updating Node.js 10 function groupEvents(us-central1)...
i  functions: updating Node.js 10 function addUserToGroup(us-central1)...
i  functions: updating Node.js 10 function unsubscribeUserFromTopicOnDelete(us-central1)...
i  functions: updating Node.js 10 function participantRemoved(us-central1)...
i  functions: updating Node.js 10 function groupAdminChanged(us-central1)...
i  functions: updating Node.js 10 function groupInfoChanged(us-central1)...
i  functions: updating Node.js 10 function deleteMessageForGroup(us-central1)...
i  functions: updating Node.js 10 function deleteMessageForBroadcast(us-central1)...
i  functions: updating Node.js 10 function deleteMessage(us-central1)...
i  functions: updating Node.js 10 function saveUidOnLogin(us-central1)...
i  functions: updating Node.js 10 function getTime(us-central1)...
i  functions: updating Node.js 10 function setStatusCount(us-central1)...
i  functions: updating Node.js 10 function subscribeToBroadcast(us-central1)...
i  functions: updating Node.js 10 function resubscribeUserToBroadcasts(us-central1)...
i  functions: updating Node.js 10 function sendMessageToBroadcast(us-central1)...
i  functions: updating Node.js 10 function unsubscribeUserFromBroadcast(us-central1)...
i  functions: updating Node.js 10 function sendUnDeliveredNotifications(us-central1)...
i  functions: updating Node.js 10 function sendNewCallNotification(us-central1)...
i  functions: updating Node.js 10 function indexNewCall(us-central1)...
i  functions: updating Node.js 10 function indexNewGroupCall(us-central1)...
i  functions: updating Node.js 10 function indexPKToken(us-central1)...
i  functions: updating Node.js 10 function deviceIdChanged(us-central1)...
i  functions: updating Node.js 10 function getVirgilJwt(us-central1)...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/subscribeToBroadcast. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/setStatusCount. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/groupInfoChanged. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/participantAdded. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/indexNewGroupCall. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/getVirgilJwt. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/deleteMessageForGroup. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/unsubscribeUserFromBroadcast. Waiting to retry...
i  functions: updating Node.js 10 function subscribeToBroadcast(us-central1)...
i  functions: updating Node.js 10 function setStatusCount(us-central1)...
i  functions: updating Node.js 10 function groupInfoChanged(us-central1)...
i  functions: updating Node.js 10 function participantAdded(us-central1)...
i  functions: updating Node.js 10 function indexNewGroupCall(us-central1)...
i  functions: updating Node.js 10 function getVirgilJwt(us-central1)...
i  functions: updating Node.js 10 function deleteMessageForGroup(us-central1)...
i  functions: updating Node.js 10 function unsubscribeUserFromBroadcast(us-central1)...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/groupInfoChanged. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/participantAdded. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/setStatusCount. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/getVirgilJwt. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/subscribeToBroadcast. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/deleteMessageForGroup. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/indexNewGroupCall. Waiting to retry...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/unsubscribeUserFromBroadcast. Waiting to retry...
i  functions: updating Node.js 10 function groupInfoChanged(us-central1)...
i  functions: updating Node.js 10 function participantAdded(us-central1)...
i  functions: updating Node.js 10 function setStatusCount(us-central1)...
i  functions: updating Node.js 10 function getVirgilJwt(us-central1)...
i  functions: updating Node.js 10 function subscribeToBroadcast(us-central1)...
i  functions: updating Node.js 10 function deleteMessageForGroup(us-central1)...
i  functions: updating Node.js 10 function indexNewGroupCall(us-central1)...
i  functions: updating Node.js 10 function unsubscribeUserFromBroadcast(us-central1)...
⚠  functions: got "Quota Exceeded" error while trying to update projects/ichat-daac0/locations/us-central1/functions/indexNewGroupCall. Waiting to retry...
i  functions: updating Node.js 10 function indexNewGroupCall(us-central1)...

Functions deploy had errors with the following functions:
	sendMessageNotification(us-central1)
	saveUidOnLogin(us-central1)
	groupEvents(us-central1)
	sendNewCallNotification(us-central1)
	indexNewCall(us-central1)
	addUserToGroup(us-central1)
	resubscribeUserToBroadcasts(us-central1)
	deviceIdChanged(us-central1)
	deleteMessage(us-central1)
	sendUnDeliveredNotifications(us-central1)
	sendMessagesForGroups(us-central1)
	deleteMessageForBroadcast(us-central1)
	unsubscribeUserFromTopicOnDelete(us-central1)
	indexPKToken(us-central1)
	sendMessageToBroadcast(us-central1)
	groupAdminChanged(us-central1)
	getTime(us-central1)
	participantRemoved(us-central1)
	participantAdded(us-central1)
	subscribeToBroadcast(us-central1)
	groupInfoChanged(us-central1)
	getVirgilJwt(us-central1)
	unsubscribeUserFromBroadcast(us-central1)
	deleteMessageForGroup(us-central1)
	setStatusCount(us-central1)
	indexNewGroupCall(us-central1)

To try redeploying those functions, run:
    firebase deploy --only "functions:sendMessageNotification,functions:saveUidOnLogin,functions:groupEvents,functions:sendNewCallNotification,functions:indexNewCall,functions:addUserToGroup,functions:resubscribeUserToBroadcasts,functions:deviceIdChanged,functions:deleteMessage,functions:sendUnDeliveredNotifications,functions:sendMessagesForGroups,functions:deleteMessageForBroadcast,functions:unsubscribeUserFromTopicOnDelete,functions:indexPKToken,functions:sendMessageToBroadcast,functions:groupAdminChanged,functions:getTime,functions:participantRemoved,functions:participantAdded,functions:subscribeToBroadcast,functions:groupInfoChanged,functions:getVirgilJwt,functions:unsubscribeUserFromBroadcast,functions:deleteMessageForGroup,functions:setStatusCount,functions:indexNewGroupCall"

To continue deploying other features (such as database), run:
    firebase deploy --except functions

Error: Functions did not deploy properly.
mac@bogon functions % 

```

**Comment by milo-** [comment]
Thi is also happening for us pretty consistently, 3-5 will fail with quota exceeded when deploying 60 functions. We're using firebase-tools 9.23.0.

**Comment by joehan** [comment]
Hey all - looks like there are a few different issues here, so I'll take them one at a time:
- To the original issue of deployments of a large number of functions hitting quota errors and failing:  As @abeisgoat mentioned, we are aware of the problems encountered when deploying lots of functions at once. We're currently working on some improvements to the throttler logic that ought to help - however, we strongly recommend making small groups of functions and deploying them in smaller groups. https://firebase.google.com/docs/functions/organize-functions#group_functions has a good guide on how to do this.
- To the `i functions: cleaning up build files... Cleaning up project directory and file based variables 00:01 ERROR: Job failed: exit code 1` error message: This message doesn't affect the success or failure of your function deployments - after a deployment is finished, the CLI tries to clean up some build artifacts from Google Container Registry to save some storage space.
- @yendiscart It looks like you are experiencing a different issue than the one described here. Could you retry your deployment with the --debug flag (`firebase deploy --only "functions:resubscribeUserToBroadcasts" --debug` and open a new issue including the output of that command?

**Comment by ggirotto** [comment]
@joehan The original problem that relates with the large number of functions is the same as I'm facing? Because I have a similar issue that outputs a quota error although I'm trying to deploy around 20 functions only. I'm facing:

`{"@type":"type.googleapis.com/google.cloud.audit.AuditLog","status":{"code":8,"message":"Build failed: Quota exceeded for quota metric 'Build Create requests' and limit 'Build Create requests per minute' of service 'cloudbuild.googleapis.com' for consumer 'project_number:<NUMBER>'."},"authenticationInfo":{"principalEmail":"<EMAIL>"},"serviceName":"cloudfunctions.googleapis.com","methodName":"google.cloud.functions.v1.CloudFunctionsService.UpdateFunction","resourceName":"projects/ticker-dev-4d59e/locations/southamerica-east1/functions/<FUNCTION_NAME>"} 
`

**Comment by inlined** [comment]
For clarification, are these failures longstanding issues or regressions as of a recent version of `firebase-tools`? We did some major refactoring recently and it's possible we have a regression for large codebases

**Comment by yendiscart** [comment]
My error was solved by upgrading my billing or plan to Blaze before I could deploy functions.

> HOW do you solve this issue? Tried to deploy one by one and still getting errors.
> 
> To try redeploying those functions, run: firebase deploy --only "functions:sendMessageNotification,functions:groupAdminChanged,functions:deleteMessageForGroup,functions:setStatusCount,functions:groupInfoChanged,functions:deleteMessageForBroadcast,functions:unsubscribeUserFromBroadcast,functions:unsubscribeUserFromTopicOnDelete,functions:saveUidOnLogin,functions:indexNewGroupCall,functions:sendMessagesForGroups,functions:addUserToGroup,functions:participantRemoved,functions:participantAdded,functions:groupEvents,functions:deleteMessage,functions:getVirgilJwt,functions:sendMessageToBroadcast,functions:indexPKToken,functions:sendNewCallNotification,functions:subscribeToBroadcast,functions:indexNewCall,functions:sendUnDeliveredNotifications,functions:deviceIdChanged,functions:resubscribeUserToBroadcasts,functions:getTime"
> 
> To continue deploying other features (such as database), run: firebase deploy --except functions
> 
> Error: Functions did not deploy properly. mac@bogon functions % firebase deploy --only "functions:getTime"
> 
> === Deploying to 'ichat-daac0'...
> 
> i deploying functions i functions: ensuring required API cloudfunctions.googleapis.com is enabled... i functions: ensuring required API cloudbuild.googleapis.com is enabled... ✔ functions: required API cloudbuild.googleapis.com is enabled ✔ functions: required API cloudfunctions.googleapis.com is enabled i functions: preparing functions directory for uploading... i functions: packaged functions (73.93 KB) for uploading ✔ functions: functions folder uploaded successfully i functions: updating Node.js 10 function getTime(us-central1)...
> 
> Functions deploy had errors with the following functions: getTime(us-central1)
> 
> To try redeploying those functions, run: firebase deploy --only "functions:getTime"
> 
> To continue deploying other features (such as database), run: firebase deploy --except functions
> 
> Error: Functions did not deploy properly. mac@bogon functions % functions % firebase deploy --only "functions:resubscribeUserToBroadcasts" mac@bogon functions % firebase deploy --only "functions:resubscribeUserToBroadcasts"
> 
> === Deploying to 'ichat-daac0'...
> 
> i deploying functions i functions: ensuring required API cloudfunctions.googleapis.com is enabled... i functions: ensuring required API cloudbuild.googleapis.com is enabled... ✔ functions: required API cloudfunctions.googleapis.com is enabled ✔ functions: required API cloudbuild.googleapis.com is enabled i functions: preparing functions directory for uploading... i functions: packaged functions (73.93 KB) for uploading ✔ functions: functions folder uploaded successfully i functions: updating Node.js 10 function resubscribeUserToBroadcasts(us-central1)...
> 
> Functions deploy had errors with the following functions: resubscribeUserToBroadcasts(us-central1)
> 
> To try redeploying those functions, run: firebase deploy --only "functions:resubscribeUserToBroadcasts"
> 
> To continue deploying other features (such as database), run: firebase deploy --except functions
> 
> Error: Functions did not deploy properly.
> 
> Having trouble? Try firebase [command] --help

**Comment by inlined** [comment]
Will close soon if nobody is still experiencing issues.

**Comment by 42ae** [comment]
Unfortunately I am also experiencing these issue often (if not each time) trying to deploy all my GCF at once

`⚠ functions: got "Quota Exceeded" error while trying to update *********** Waiting to retry...`

I know it's best to deploy only the group of functions that have been updated and leave the other ones untouched, but sometimes when updating environment configuration it is required to deploy all cloud function to allow the new configuration to propagate properly.

Please advise on this issue, it would be great to find a better workaround to allow better integration with CI too.

Thanks!

**Comment by aurimas-fertilemind** [comment]
Experiencing same issue while deploying 62 functions.
```
⚠  functions: got "Quota Exceeded" error while trying to update projects/<redacted>/locations/us-central1/functions/<redacted>. Waiting to retry...
```

2 functions fails to deploy, thus I assume there should be somewhere 60 function deployment limit?

As @42ae mentioned previously, sometimes bulk deployment is needed in order for all functions to be updated. Looking forward hearing for solutions to this problem.

**Comment by vishen** [comment]
We are also experiencing something similar with about 80 functions where 5-6 regularly fail. We are currently on version `10.1.4`.

```
functions: got "Quota Exceeded" error while trying to update projects/***/locations/northamerica-northeast1/functions/<function>. Waiting to retry...
```

**Comment by trex-quo** [comment]
I am frequently experiencing this issue as well while deploying over 100 functions. We deploy in groups and have created some parsing logic for our CI system, however it would be very convenient to have these auto-retry if we pass the deployment quota, or at least spit out a command after failure to deploy the rest of the functions (currently it just lists them). 

The current text says "waiting to retry" while we exceed the quota, so that was the behavior I was expecting, however it is not what ended up happening.

**Comment by codin-play** [comment]
I can consistently get this error even when deploying a group of only 5 functions. The error message literally shows nothing useful and it has been a real pain. I can deploy each one individually which results in deployments for a full stack to take hours. Any ideas from folks on how to overcome this would be really appreciated.

```
$ firebase --version
10.5.0

$ gcloud --version
Google Cloud SDK 369.0.0
alpha 2022.01.14
beta 2022.01.14
bq 2.0.72
core 2022.01.14
gsutil 5.6

```

```
Deployment failure:
Build failed: {"metrics":{},"error":{"buildpackId":"","buildpackVersion":"","errorType":"OK","canonicalCode":"OK","errorId":"","errorMessage":""},"stats":[{"buildpackId":"google.utils.archive-source","buildpackVersion":"0.0.1","totalDurationMs":199,"userDurationMs":198},{"buildpackId":"google.nodejs.npm","buildpackVersion":"0.9.0","totalDurationMs":19905,"userDurationMs":19841},{"buildpackId":"google.nodejs.functions-framework","buildpackVersion":"0.9.4","totalDurationMs":3764,"userDurationMs":3744},{"buildpackId":"google.utils.label","buildpackVersion":"0.0.2","totalDurationMs":0,"userDurationMs":0}],"warnings":null,"customImage":false}
```

**Comment by inlined** [comment]
We will prioritize looking into this soon. There will be a three pronged strategy:

1. Fixing the "invalid source token" problem, which occurs when function deploys take a very long time
2. Looking at tuning the rate limiter to avoid quota errors
3. Improving deploy code so that it can recognize and skip functions that don't need redeployment

Hopefully this will help you all when we can get these improvements ready.

**Comment by trex-quo** [comment]
This is encouraging to hear; thank you for the update!

**Comment by olve** [comment]
> We will prioritize looking into this soon. 

any updates on this?

**Comment by inlined** [comment]
We've currently launched (but not well documented or added wizzard support for) codebases. With codebases you can have multiple directories of Cloud Functions. Simply create a firebase.json fragment that looks a bit like:

```json
"functions" : [
  {
    "source": "functions",
    "codebase": "main"
  }, {
    "source": "otherfunctions",
    "codebase": "other"
  }
]
```

With this, you can now call `firebase deploy functions:main` or `firebase deploy functions:other`. Next, we're working on recognizing that a codebase hasn't been modified since last deployment and we'll skip those functions altogether.

**Comment by sluramod** [comment]
@inlined is it possible there is a regression and https://github.com/firebase/firebase-tools/issues/2606 / https://github.com/firebase/firebase-tools/pull/3246 no longer works? 
Is it possible to restore 'retry on quota' functionality?

**Comment by gustavopch** [comment]
I'm getting "Invaild source token" every time for the same 2 functions when deploying 24 Node.js 16 functions at once. If I deploy then individually, there's no error. I'm using `firebase-tools@11.1.0`.

**Comment by gustavopch** [comment]
So, I was trying to deploy 24 functions at once and 2 failed. I've experimented more and realized that if I deploy exactly those 2 problematic functions, deployment happens just fine. If I deploy the 2 problematic functions + another function, the 2 problematic functions fail again while the other function is successfully deployed.

**Comment by inlined** [comment]
Scanning through some bugs and came across this one. I wanted to give a few updates:
1. **Regression/Fix**. I'm glad that the regression seems fixed already and sorry that we had a regression.
2. **Invalid source token**. We have not started a fix here. If someone wants to try fixing this, it requires some cleverness but not knowledge of the codebase
3. **Quotas** You may find some (seemingly) contradictory advice about running w/ --only and groups of 10 functions or that we may run your functions in batches. The Firebase CLI already batches your functions for deploy thanks to the ThrottlerQueue class; this is basically a JS equivalent of a threadpool and makes sure we only have N functions deploying concurrently. Then with backoffs we approximate the most calls you can make in a project within quota limits. It's a pipe dream to have an improved request queue that is quota aware so it can do smarter backoffs, but we may not get there for a long time. Manually deploying only a few functions at a time is even more conservative than normal. You are _not_ supposed to run that command several times in parallel, since that removes all benefits of fewer deploys. In the future, Codebases will also help you limit the number of functions that are deployed.
4. **Codebases feature**

> Wouldn't it be better if there was an option to deploy only modified functions? (--only-changed?) Usually a single push only affects a few functions so deploying only those would help.

@Kopunk Unfortunately, a more accurate implementation of `--only-changed` that _isn't_ the codebases feature would be about as complicated as writing a JavaScript interpreter (which means that a true implementation of --only-changed is probably also impossible due to the Halting Problem). Any non-trivial cloud function leans on utility methods. Any change to any one of these utility methods will change some amount of code that can only be determined by running through a program.

Codebases is a compromise. If you agree to put firewalls between parts of your code (e.g. directories) we can instead do a file diff and notice nothing changes. Sometimes you may have shared code and it truly isn't possible to separate your code cleanly (e.g. there are shared interfaces). For the first several years of its existence, Firebase solved this problem using an internal-API node module. I've heard that Yarn workspaces can solve this quite eloquently as well. Many teams at Google use them.

**Comment by willhackett** [comment]
We've got a large number of functions (~150) we deploy regularly. We get maybe 1 - 2 failed function deploys per run due to the "Quota Exceeded" error. 

I thought `firebase-tools` was meant to manage this & retry, but we're not seeing this logic succeed.

Is there any way to work around this without having to rearchitect our entire production application? I understand the codebases pattern, but this won't work for our use case given all the functions are camel-case named.

**Comment by taeold** [comment]
@willhackett I'd love to understand what you mean by camlcased name functions better  - can you elaborate why codebases doesn't fit your application?

I'm planning to review https://github.com/firebase/firebase-tools/pull/4967 soon, but my feeling is that fighting the GCF quota limit isn't fool-proof way, especially since quota is different depending on your project's state, location, history, etc.

Our thesis is that codebase with feature to automcatically skip deployment of function that doesn't require a re-deploy is more promising and I'd love to understand in what ways we could improve codebase support to work more developers' setups.

**Comment by gustavopch** [comment]
@taeold Suppose I want to have very granular updates. If a single function changed, I want to redeploy just that one function and no other. Would it make sense to have 1 function per codebase? Any possible issues? Or do you think that would be over-engineering in the first place?

**Comment by taeold** [comment]
@gustavopch I think somewhere between the 2 extremes (1 codebase 100 functions vs 100 codebase 1 function) would help. For example, you might group function in a codebase by trigger type. Or by their business domain. In the ideal case, when you are apply a bug fix or a new feature, it would affect a single codebase. But I suspect that our reality is a bit messier than that.

You can give "skip deploying a codebase that's doesn't require one" right now by using the latest Firebase CLI and enabling the priview:

```
firebase --open-sesame skipdeployingnoopfunctions
```

We are making some last pushes to get this feature cleaned up, an when its ready you see it be mentioned in our documentations. Would love to hear from advanced users like yourselves what you think.

**Comment by gustavopch** [comment]
@gustavopch Oh, interesting. Currently, I have a script that uses ESBuild to bundle each function independently and generate a hash, so when I'm deploying, I only deploy functions whose hashes changed. Is `skipdeployingnoopfunctions` similar to that? Can you explain what logic it performs to detect what to deploy and what not?

**Comment by taeold** [comment]
@gustavopch Nice - I've been experimenting with esbuild too and liked the results.

Yes, very much similar to what you described. In addition, we check that the secrets associated with your function - if there is a new version of a secret linked to your function, we mark it as re-deploy.

**Comment by TheIronDev** [comment]
`skipdeployingnoopfunctions` generates a hash from:

* the contents of the files included in each codebase.
* Runtime Configuration 
* Secret Versions
* Environment Variables

If a file in a codebase changes (add a space, etc.), then the generated hash for that file (and as a result the subsequent codebase) will receive a new hash.

**Comment by gustavopch** [comment]
@TheIronDev @taeold Thanks for clarifying.

> I think somewhere between the 2 extremes (1 codebase 100 functions vs 100 codebase 1 function) would help.

What problems could I find if I put each function in a separate codebase?

**Comment by taeold** [comment]
@gustavopch That's an interesting stress case! One edge case you might consider is that a Cloud Function deploy is actually a 2 step process:

1) Create container image from your source code
2) Deploy your container image to GCF

Since we don't want to create the same container image for every function you have, we have special logic to generate 1 container image per codebase* and to re-use it for all functions in the codebase.

So if you have 100 functions in 1 codebase, your deploy will call out to Cloud Build once to generate and reuse a single container image. If you have 100 codebases with 1 functions each, you will call out to Cloud Build 100 times. That probably is going to hit your Cloud Build quota in some way.


* - I'm lying here a bit. We actually generate 1 container image per codebase-memory pair for technical reasons.

**Comment by gustavopch** [comment]
@taeold If I have 100 functions (each in an isolated codebase) and deploy a change that affects only 5 functions, will it consume quota for 100 containers or for 5 containers?

**Comment by s11richard** [comment]
We have about 97 functions, and consistently get Quota errors when deploying them, which then requires multiple retries to fix our deployment pipeline. This problem obviously gets worse as we add more functions.

It would be great if there was some guidance on how to manage a project that has 100+ functions. Or is that not supported/expected?

It seems like there used to be some retry capabilities. Having those reinstated would be very helpful. Is that on the roadmap at all? Is any other fix to this issue?

**Comment by hefgi** [comment]
> It would be great if there was some guidance on how to manage a project that has 100+ functions.

**if anyone could confirm this please!**

From my understanding, the best practices for a scalable project, is to "group" the functions in folders or repositories (or codebases). The idea is to have < 30 functions per "group" in order to avoid the `"Quota Exceeded"` problem.

Here's the doc from Firebase: https://firebase.google.com/docs/functions/organize-functions
Here's a tutorial: https://codeburst.io/organizing-your-firebase-cloud-functions-67dc17b3b0da

**Comment by taeold** [comment]
@s11richard @hefgi Thanks for reaching out!

While I can see improvements in trying to improve experience in deploying large amount of functions with a single deploy command, it's always going to be hard because Google Cloud Platform's quota is going to be dynamic and different for every user. We also noticed that it often changes over time and can be different for each project for the same user.

Our current function deploy impl. is an attempt to navigate the quota as best as we could, and while I'm pretty sure that it's an improvement over our past impl., it's obvious from this issue that we still have much to go to cover everyone's use cases.

Lately, we've been experimenting from another direction:

> Is there a way to reduce the number of functions that's deployed on each deployment?

Say make a code change that affects runtime behavior of a subset of your functions. Do we really need to deploy all of the functions?

@hefgi's suggestion of breaking your functions into codebases is part of an answer. But it still means you need to manually pick which codebases to deploy on each run.

@TheIronDev recently worked on a new feature to automatically detect and skip deployment of functions in a codebase without any changes.

I'd strongly recommend you give codebases and "no-op deploy" a try. In the coming weeks, we'll try to publish more public guidance on how we think we can improve reliability of deploys for projects with many many functions.

**Comment by dtran320** [comment]
Thanks so much for all the advice @taeold. 

Unfortunately for us, refactoring to move to codebases isn't feasible at the moment. 

Looking at the [changelog for firebase-functions 11.13.0](https://github.com/firebase/firebase-tools/releases/tag/v11.13.0), could you help clarify what[ the change from #5032's flipping of the `skip no-op deploys` flag to `true` does](https://github.com/firebase/firebase-tools/pull/5032/files)? Does that ONLY apply to diff-ing the hashes for entire codebases and only applying a skip no-op deploy for a set of codebase functions, or is it meant to hash each individual function and skip the no-op function deploy for each individual function? We updated our fork of the [firebase-action for Github](https://github.com/Flow-Club/firebase-action/blob/master/Dockerfile) to use `firebase-tools 11.13.0` but I believe I'm still seeing it deploy every function even though they should all be no-ops. Thanks so much!

**Comment by TheIronDev** [comment]
@dtran320 

> Looking at the changelog for firebase-functions 11.13.0, could you help clarify what the change from #5032's flipping of the skip no-op deploys flag to true does?

That enables the no-op deploy detection by default.

> Does that ONLY apply to diff-ing the hashes for entire codebases and only applying a skip no-op deploy for a set of codebase functions, or is it meant to hash each individual function and skip the no-op function deploy for each individual function?

Applying skip deployment of no-op deploys is applied to a set of functions within a codebase.

> We updated our fork of the firebase-action for Github to use firebase-tools 11.13.0 but I believe I'm still seeing it deploy every function even though they should all be no-ops. 

That shouldn't be the case..... diffing codebases occur by checking:

1. the codebase's folder
1. environment variables
1. secret versions

If you want to check, I'd suggest doing thee following from a single box

1. make sure you're on > firebase-tools@11.13.0
1. run `firebase deploy` 
1. run `firebase deploy` again (from the same box)

You *should* see:

* `functions: Skipping the deploy of unchanged functions with experimental support for skipdeployingnoopfunctions`
*  `functions[ ... ] Skipped (No changes detected)`

**Comment by gustavopch** [comment]
> 



> If I have 100 functions (each in an isolated codebase) and deploy a change that affects only 5 functions, will it consume quota for 100 containers or for 5 containers?

@TheIronDev What about this?

**Comment by dtran320** [comment]
Thanks @taeold!
>If you want to check, I'd suggest doing thee following from a single box

> make sure you're on > firebase-tools@11.13.0
> run firebase deploy
> run firebase deploy again (from the same box)

Thought this might be an issue with our CI, so deployed locally several times in a row (without changes), and when they finally go through without the quota error, still haven't been able to get anything to successfully display `Skipping the deploy of unchanged functions...`

```sh
% firebase --version
11.13.0
```

```sh
firebase deploy --only functions

=== Deploying to <REDACTED>...

i  deploying functions
✔  functions: Finished running predeploy script.
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
i  functions: preparing codebase default for deployment
i  functions: preparing functions directory for uploading...
i  functions: packaged /path/to/functions (7.41 MB) for uploading
i  functions: ensuring required API cloudscheduler.googleapis.com is enabled...
✔  functions: required API cloudscheduler.googleapis.com is enabled
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 14 function <REDACTED>(us-central1)...
i  functions: updating Node.js 14 function <REDACTED>(us-central1)...
...
✔  functions[<REDACTED>(us-central1)] Successful update operation.
✔  functions[<REDACTED>(us-central1)] Successful update operation.
...
Function URL (<REDACTED>(us-central1)): https://<REDACTED>
Function URL (<REDACTED>(us-central1)): https://<REDACTED>
...
i  functions: cleaning up build files...

✔  Deploy complete!
```

Nothing changes in subsequent deploys

```sh
firebase deploy --only functions

=== Deploying to <REDACTED>...

i  deploying functions
✔  functions: Finished running predeploy script.
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
i  functions: preparing codebase default for deployment
i  functions: preparing functions directory for uploading...
i  functions: packaged /path/to/functions (7.41 MB) for uploading
i  functions: ensuring required API cloudscheduler.googleapis.com is enabled...
✔  functions: required API cloudscheduler.googleapis.com is enabled
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 14 function <REDACTED>(us-central1)...
i  functions: updating Node.js 14 function <REDACTED>(us-central1)...
...
✔  functions[<REDACTED>(us-central1)] Successful update operation.
✔  functions[<REDACTED>(us-central1)] Successful update operation.
...
Function URL (<REDACTED>(us-central1)): https://<REDACTED>
Function URL (<REDACTED>(us-central1)): https://<REDACTED>
...
i  functions: cleaning up build files...

✔  Deploy complete!
```

Is the build cleanup potentially an issue? If anyone else has gotten this to work, tips would be greatly appreciated!

**Comment by mike-4040** [comment]
We have about 150 functions.
```
% firebase -V
11.14.0
```
During the first deploy, 33 functions failed.
Second attempt: seen some `Skipped (No changes detected)`, but only a handful (10).
Mostly firebase was trying to deploy the same function again even after a successful deployment.

**Comment by alexlouden** [comment]
We have just over 100 functions, and similarly have had a lot of issues with deploy reliability/quota issues (including exhausting the daily quota limit). We are currently using a custom deploy script to deploy them in batches of 20 functions at a time using `--only` - this was the only way to make continuous deployments reliable, however deploys take 30~45 minutes. (If there's interest I can tidy up and share this script!)

I've updated to `11.14.0`, re-run the deployment on our CI server twice (no filesystem/env changes) and I'm not seeing any skipped functions - all are re-deploying. We're not using codebases at the moment - just a single "default" codebase.

The two things that are preventing us from migrating to codebases are:
1. We have some common modules/shared code - for managing config, firestore setup and getters/setters, firebase auth helpers, internal utils. We could split these out using yarn workspaces as mentioned above, but this adds extra developer overhead and complexity for testing/mocking/etc.
  I'd prefer a solution that just re-deployed all functions if a specific folder of shared code changed, or ideally operated on a per-function basis - e.g. by hashing each bundled & tree-shaken function? (the ESBuild approach mentioned above sounds great to me!)
  
2. Since the names of HTTP functions form the public function URL (`https://us-central1-myproject.cloudfunctions.net/myFunctionName`), we can't rename these functions since the URLs are used by external services and in native client apps. (We now give new HTTP functions nice URLs using firebase hosting rewrites, but that doesn't help with existing URLs that are in use).
  If we had control over the function URLs or if we could maintain the current URLs (`/myFunctionName` rather than `/mycodebase-myFunctionName`) then we could get around this.

**Comment by alexlouden** [comment]
I was chatting to some friends who use AWS - looks like lambda uses the esbuild approach: https://github.com/aws/aws-cdk/tree/main/packages/%40aws-cdk/aws-lambda-nodejs 
Could be useful for reference ☺️

**Comment by gustavopch** [comment]
On a high level:
- Identify the functions that are exported by the entry files.
- For each function, create a temporary file that imports it.
- Pass that file as an entry point to ESBuild with `bundle: true` and `write: false`.
- Use the result to generate a hash.

The problem is how to identify what functions need to be redeployed when dependencies in `package.json` change. Ideally, bumping a dependency that's used by a single function shouldn't cause all the other functions to be redeployed. Maybe use regex to identify the imports/requires in the output produced by ESBuild.

**Comment by gustavopch** [comment]
I'd be happy enough if somehow I could inform `firebase-tools` about the hashes that I generated by myself.
```sh
firebase deploy hashes:{"hosting":{"mysite":"abc123"},"functions":{"fn1":"bcd234","fn2":"cde345"}}
```

**Comment by taeold** [comment]
@alexlouden Thanks for an awesome writeup. The migration pains you described are really insightful and I'll take this back to the team.

Codebase is just a start and we want to continue to explore we can better support projects with many many function. It sounds ESBuild is an interesting approach we can explore - we've also had internal conversation about whether natively supporting `npm` or `yarn` workspaces would help with folk's need.

One other direction we've brainstormed (specifically with v2 functions) is an idea of "eventing monolith" - i.e. a single Cloud Function instance that's responsible for processing multiple event sources. Consolidating functions that are called sparesley would reduce number of cold start and possibly save on infra cost.


I'll go back to the team to look into skip-deployment issue. Thanks for the heads up.

**Comment by alexlouden** [comment]
Can we reopen please? @google-oss-bot

**Comment by alexlouden** [comment]
Bad bot. Can we reopen please? 😀

**Comment by Splaktar** [comment]
> * skipdeployingnoopfunctions

How do I disable this new default behavior?

**Comment by TheIronDev** [comment]
> > * skipdeployingnoopfunctions
> 
> How do I disable this new default behavior?

There isn't a feature to disable the behavior. (but that could be a worthwhile feature request to gather public opinion.)

That said, if you invoke deploy on a function with the `--only` flag, you force deployment.

Example:

```
firebase deploy --only functions:makeUppercase
```

^ This will forcefully deploy the `makeUppercase` function.

[More info on the --only deploy](https://firebase.google.com/docs/functions/manage-functions#deploy_functions)

**Comment by aryzle** [comment]
@taeold @inlined can we reopen pls? and any updates or recommended best practices here?

Splitting up the codebase to take advantage of "codebases" is really clunky, especially when a team is just doing it for deploy optimization... leads to having to manage multiple package.json and yarn.lock files, changing how you build things, etc.

**Comment by gustavopch** [comment]
> One other direction we've brainstormed (specifically with v2 functions) is an idea of "eventing monolith" - i.e. a single Cloud Function instance that's responsible for processing multiple event sources. Consolidating functions that are called sparesley would reduce number of cold start and possibly save on infra cost.

Hey @taeold, did the team discuss this idea any further? 🥺🤞 It's great because it eliminates deployment complexity, saves on infra cost and improves the response times, all at once.

**Comment by pedrofsantoscom** [comment]
> @dtran320
> ...
> 
> If you want to check, I'd suggest doing thee following from a single box
> 
> 1. make sure you're on > firebase-tools@11.13.0
> 2. run `firebase deploy`
> 3. run `firebase deploy` again (from the same box)
> 
> You _should_ see:
> 
> * `functions: Skipping the deploy of unchanged functions with experimental support for skipdeployingnoopfunctions`
> * `functions[ ... ] Skipped (No changes detected)`

I only want to deploy cloud functions so I need to run `firebase deploy --only functions`, and this way the `skipdeployingnoopfunctions` doesn't run/work.

**Comment by mercury2269** [comment]
> I only want to deploy cloud functions so I need to run firebase deploy --only functions, and this way the skipdeployingnoopfunctions doesn't run/work.

@pedrofsantoscom This was not working for me either, after digging through the code I found that the issue was with runtime configuration producing a new hash every time. I have a [PR fix](https://github.com/firebase/firebase-tools/pull/5355) for this open.

**Comment by mvergarair** [comment]
@pedrofsantoscom does firebase deploy skip non-changed functions for you? I'm using the latest version and it is not working.

**Comment by wobsoriano** [comment]
@pedrofsantoscom same here with latest version.

**Comment by wobsoriano** [comment]
@mvergarair same here with latest version.

**Comment by pedrofsantoscom** [comment]
> @pedrofsantoscom does `firebase deploy` skip non-changed functions for you? I'm using the latest version and it is not working.

I haven't tested that as it will deploy everything, which i don't want.

**Comment by pedrofsantoscom** [comment]
> > I only want to deploy cloud functions so I need to run firebase deploy --only functions, and this way the skipdeployingnoopfunctions doesn't run/work.
> 
> @pedrofsantoscom This was not working for me either, after digging through the code I found that the issue was with runtime configuration producing a new hash every time. I have a [PR fix](https://github.com/firebase/firebase-tools/pull/5355) for this open.

[your fix is deployed!](https://github.com/firebase/firebase-tools/releases/tag/v11.21.0#:~:text=Fixes%20an%20issue%20where%20already%20deployed%20functions%20with%20the%20same%20remote%20configuration%20do%20not%20get%20skipped%20(%235354)) 🚀

**Comment by edenhermelin** [comment]
just updated to `v11.21.0` and the entire deployment failes 😞 
looks like a regression, at least for monorepos with `pnpm`.
Relevant issue seems to be this one:
https://github.com/firebase/firebase-tools/issues/4952#issue-1362289728

**Comment by sulman2** [comment]
So updating to the latest version of firebase CLI worked for me now it is saying `Skipping the deploy of unchanged functions with experimental support for skipdeployingnoopfunctions` and it skips as well which is great. But the whole point of this functionality is if I have 100 functions and I make changes to 1 function then running the deploy command will skip 99 functions and deploy only the modified function.

So what happened is
1. I ran the command `firebase deploy --only functions` the first time after updating CLI and it deployed all the functions (good)
2. I ran the command again and it skipped all the functions. (Great)
3. I made a change to a function and ran the command again and it started deploying all the functions again (bad)

What am I doing wrong?

**Comment by NNskelly** [comment]
skipdeployingnoopfunctions seems to be tripping over itself.  As of firebase-tools 11.29.1,
- changing even 1 function seems to sometimes redeploy all functions with `--only functions`
- changing 2 functions and doing a general `deploy --only functions` redeployed all (~30) of my functions, and failed to update 9 which were previously working and unchanged.
- after the failure, attempting to `deploy --only functions` again to resolve errors _skips_ every single function under the excuse of skipdeployingnoopfunctions.
So, following the thread, the only way out of this situation, since there's no way to force-disable the skip-noop behavior, is to visually track which functions randomly fail on a given deployment and re-run them all individually with `deploy --only functions:IndividualFailedFunctionName`?

**Comment by gustavopch** [comment]
> One other direction we've brainstormed (specifically with v2 functions) is an idea of "eventing monolith" - i.e. a single Cloud Function instance that's responsible for processing multiple event sources. Consolidating functions that are called sparesley would reduce number of cold start and possibly save on infra cost.

@taeold Any progress on this idea?

**Comment by NNskelly** [comment]
I have since learned that there is a poory/un-documented syntax for manual-deploying some-but-not-all functions:
`firebase deploy --only functions:myFunc1,functions:myFunc2,functions:myFunc3`
Not a fix, but a very useful stopgap when an arbitrary 5 of your 50 functions fail to deploy for no actionable reason.

**Comment by ottob** [comment]
These errors are appearing much more often than the used to:
Failed to configure trigger for event-type:providers/cloud.firestore/eventTypes/document.write resource:projects/ourproject/databases/(default)/documents/payments/{id} service:firestore.googleapis.com.

Is it because they are 1gen functions?
Using the [workaround](https://github.com/firebase/firebase-tools/issues/3919#issuecomment-1595112552) above not really practical if 20/100 functions failed to deploy.

Please try to fix the root cause and please perhaps add a --force option so all functions can be re-deployed even if now changes are detected.

**Comment by wobsoriano** [comment]
> I have since learned that there is a poory/un-documented syntax for manual-deploying some-but-not-all functions: `firebase deploy --only functions:myFunc1,functions:myFunc2,functions:myFunc3` Not a fix, but a very useful stopgap when an arbitrary 5 of your 50 functions fail to deploy for no actionable reason.

I have a script that gets the updated files inside a `modules` folder (where all of my firebase functions are exported) that might be helpful to some - https://gist.github.com/wobsoriano/3634d58ead2e1f80c1dc090d3f5dd8e6

**Comment by dreinon** [comment]
Nice script! What if what changes is a method in a file outside the `modules` folder, that’s imported within a file inside the modules folder but this last file does not change? We also need to detect file dependencies changes, not only file changes, am I wrong?

Enviado desde Correo<https://go.microsoft.com/fwlink/?LinkId=550986> para Windows

De: Robert ***@***.***>
Enviado: lunes, 31 de julio de 2023 17:01
Para: ***@***.***>
CC: Daniel Reinón ***@***.***>; ***@***.***>
Asunto: Re: [firebase/firebase-tools] Can't deploy all functions consistently. Fails most of the time. (Issue #3919)


I have since learned that there is a poory/un-documented syntax for manual-deploying some-but-not-all functions: firebase deploy --only functions:myFunc1,functions:myFunc2,functions:myFunc3 Not a fix, but a very useful stopgap when an arbitrary 5 of your 50 functions fail to deploy for no actionable reason.

I have a script that gets the updated files inside a modules folder (where all of my firebase functions are exported) that might be helpful to some - https://gist.github.com/wobsoriano/3634d58ead2e1f80c1dc090d3f5dd8e6

**Comment by Zloka** [comment]
Would it be possible to re-open this issue? We are running into this issue quite frequently using firebase tools v13.1.0, and deploying around 70 cloud functions.

Sometimes it works, sometimes it doesn't, and there doesn't really seem to be any other option than simply trying again.

**Comment by gbourne1** [comment]
Agree this should be reopened - almost every deploy has Quota Exceeded.

**Comment by icetosh** [comment]
Same here. I have 68 functions and usually +-10 of them fails.

And even worse, redeploy of all functions (with simple `firebase deploy`) skips ALL functions including failed.

**Comment by NNskelly** [comment]
fwiw, I have found this to be a LOT more stable in recent Firebase Tools versions (currently on 13.12.0), and after updating most functions to v2 syntax (which will be hard-required some day anyway, I suspect, although I'm still waiting on a v2 equivalent for auth triggers...).  Starting there is hopefully at least a decent mitigation.

**Comment by canbal** [comment]
I keep hitting this issues as well. This makes firebase functions not dependable when you want to setup a CI/CD with automatic deployments, and grouping functions is not an acceptable answer. Instead the deploy command should pace the deployments with whatever rate the system expects. I agree that this ticket should be opened and addressed.

**Comment by ajguyot** [comment]
Hello from 2025 where this remains an exceptionally annoying problem. I have read through this entire thread and I simply do not understand why Firebase cannot implement one of two extremely straightforward solutions:

1. Fix Firebase Tools' throttling/retry logic so it will actually just handle large function deployments automatically without erroring out on this quota. No one is asking for speed here, we just want our CI pipelines to eventually complete.
2. Just... Increase the quota? So many developers in this thread list around 100 cloud functions as their problem set. That's where my company is as well. Maybe that's a lot of functions, but this is *Google* Cloud Platform. GCP should be able to keep this quota high enough that only a very small number of outliers run into it.

Telling everyone to restructure their codebases to deal with this problem is pretty ridiculous.

**Comment by phileasthefogg** [comment]
Another bump from 2025 and Firebase-tools v14.2.2. 

This is still an issue.  Not much else to add as the problem seems to be pretty clearly outlined by all the previous comments.  Running into quota limits while working with 43 Cloud Functions.  

If a fix isn't on the immediate roadmap it would at least be helpful to know how Firebase determines which functions should be updated and which can be skipped so that we can come up with workarounds (or at least try to come up with workarounds).

**Comment by targetlucked69** [comment]
Glad we switched to Supabase + other tech and we don't have to experience the impending doom that Firebase is.

**Comment by mike-4040** [comment]
> Just wanted to highlight that this is an issue for us as well - we have to make each deployment at least twice _per environment_.

Same issue here — deployments were more reliable last year, but now they fail from time to time.
We put together a script that splits functions into batches of 30 and deploys them in chunks. It takes longer overall, but it usually works.

**Comment by alexlouden** [comment]
Our script deploys in batches of 20, so our 110+ functions take 45 minutes to deploy. Deploys are reliable now (thanks to a retry with exponential backoff if a batch fails), but the deploy speed definitely leaves room for improvement.

**Comment by BrianGrug** [comment]
Same issue here. Deplyoing ~60 cloud functions. Tried @alexlouden 's solution, but ended up running into the same issue if it wasn't timed perfectly or given a good buffer.

**Comment by davewd** [comment]
@inlined can we please re-open I hit this problem on every ci/cd deploy

**Comment by davewd** [comment]
> Our script deploys in batches of 20, so our 110+ functions take 45 minutes to deploy. Deploys are reliable now (thanks to a retry with exponential backoff if a batch fails), but the deploy speed definitely leaves room for improvement - I definitely agree it should be reopened.

@alexlouden / @BrianGrug  Do you have the script to hand please?

**Comment by Zloka** [comment]
> To solve this, deploy functions in groups of 10 or fewer.

I know the official documentation states this, but it is frankly ridiculous.

1. That is a very small batch of functions, that even hobby-projects realistically exceed.
2. It is not necessary simple to deduce which batch of functions to deploy, as good practices like code re-use might mean that changing functionality in one place, could warrant re-deploying many functions. Other cases, like updating environment variables exist, too.

If the solution is to batch deployments, why doesn't the firebase-tools handle this under the hood to begin with?

**Comment by davewd** [comment]
Thanks @Zloka , can you point me to sample yaml plz

**Comment by mike-4040** [comment]
We are deploying by batches of 30, prettier reliable. But the whole process is pain in the ass.

Firebase cli have a feature ( not sure if it’s documented) to deploy only changed functions.
Do not work for us, cause we use commit sha as release id for observability tool. And as soon as any env var changes firebase redeploy everything.

**Comment by davewd** [comment]
Anyone have sample yaml for 10 functions at a time or split by region ?

**Comment by mandalornl** [comment]
@davewd We use this deploy script in conjunction with `sywac` and `@inquirer/prompts` to manually deploy functions in batches of 20. This will most likely not work out-of-the-box for you, but at least you'll get a general idea.

<details>

<summary>
Code
</summary>

```javascript
import {
  URL,
  fileURLToPath
} from 'node:url';
import { readFile } from 'node:fs/promises';
import { join } from 'node:path';
import {
  spawn,
  execSync
} from 'node:child_process';
import { glob } from 'glob';
import {
  select,
  checkbox,
  confirm
} from '@inquirer/prompts';

const root = fileURLToPath(new URL('../../../../', import.meta.url));

/**
 * Fetch deployed functions
 *
 * @param {string} alias
 *
 * @returns {Promise<string[]>}
 */
const fetchDeployedFunctions = async (alias) => {
  const cmd = join(root, 'node_modules/.bin/firebase');

  const json = execSync(`${cmd} -P ${alias} functions:list --json`, {
    encoding: 'utf-8'
  });

  const data = JSON.parse(json);

  return data.result.map((fn) => {
    const id = fn.id.replaceAll('-', '.');

    if (
      !id.startsWith('auth.')
      && !id.startsWith('firestore.')
      && !id.startsWith('schedule.')
      && !id.startsWith('storage.')
      && !id.startsWith('task.')
      && !id.startsWith('ext.')
    ) {
      return `https.${id}`;
    }

    return id;
  });
};

/**
 * Get functions
 *
 * @param {string} alias
 *
 * @returns {Promise<{ value: string, name: undefined|string }[]>}
 */
const getFunctions = async (alias) => {
  const cwd = join(root, 'functions/src');

  const files = await glob('**/index.js', { cwd });

  const functions = await Promise.all(files.map(async (path) => {
    const segments = path.replace('/index.js', '').split('/');
    const module = await import(join(cwd, path));

    return Object.keys(module)
      .filter((key) => key !== 'default')
      .map((key) => ([
        ...segments.map((segment) => (
          segment.split('-').map((value, index) => {
            if (index === 0) {
              return value;
            }

            return `${value.charAt(0).toUpperCase()}${value.slice(1)}`;
          }).join('')
        )),
        key
      ].join('.')));
  }));

  const deployedFunctions = await fetchDeployedFunctions(alias);
  const items = functions
    .flat()
    .filter((a, index, self) => (
      !self.find((b) => b.startsWith(`${a}.`))
    ));

  const collator = new Intl.Collator();

  return [
    ...items.map((value) => {
      const isNew = !deployedFunctions.includes(value);

      return {
        value,
        name: isNew ? `${value} (new)` : value
      };
    }),
    ...deployedFunctions
      .filter((value) => !items.includes(value))
      .map((value) => {
        const disabled = value.startsWith('ext.');

        return {
          value,
          disabled,
          name: disabled ? value : `${value} (delete)`
        };
      })
  ].sort((a, b) => collator.compare(a.name ?? a.value, b.name ?? b.value));
};

/**
 * Get aliases
 *
 * @returns {Promise<{ value: string }[]>}
 */
const getAliases = async () => {
  const json = await readFile(join(root, '.firebaserc'), 'utf-8');
  const data = JSON.parse(json);

  return Object.keys(data.projects ?? {}).map((value) => ({ value }));
};

export default {
  flags: 'functions:deploy',
  group: 'Functions',
  run: async () => {
    const aliasChoices = await getAliases();
    const alias = await select({
      message: 'Alias?',
      choices: aliasChoices
    });

    const functionChoices = await getFunctions(alias);
    const functions = await checkbox({
      message: 'Functions?',
      choices: functionChoices,
      loop: false,
      validate: (values) => values.length > 0 || 'Select at least one function'
    });

    const deploy = await confirm({
      message: `Are you sure you want to deploy ${functions.length} Cloud ${functions.length === 1 ? 'Function' : 'Functions'} to '${alias}'?`,
      default: false
    });

    if (!deploy) {
      return;
    }

    do {
      const chunk = functions.splice(0, 20);
      const only = chunk.map((fn) => `functions:${fn.replace('https.', '')}`).join(',');

      let done = false;

      do {
        const child = spawn(join(root, 'node_modules/.bin/firebase'), [
          '-P',
          alias,
          'deploy',
          '--only',
          only
        ], {
          stdio: 'inherit'
        });

        try {
          await new Promise((resolve, reject) => {
            child.on('close', (code) => {
              if (code !== 0) {
                return reject(new Error(`Child process exited with code ${code}`));
              }

              resolve();
            });
          });

          done = true;
        } catch (error) {
          const retry = await confirm({
            message: 'Process failed. Retry?',
            default: true
          });

          if (!retry) {
            throw error;
          }
        }
      } while (!done);
    } while (functions.length > 0);
  }
};
```

</details>

**Comment by mike-4040** [comment]
We use bash script:

```bash
#!/usr/bin/env bash
# Deploy Firebase functions in sequential batches

set -uoe pipefail

BATCH_SIZE=30            # Number of functions per batch
SLEEP_BETWEEN_BATCHES=15 # Sleep seconds between successful batches

# Grab all cloud function paths, split off only the function names, lowercase them all
FUNCTION_NAMES=$(find ./functions/lib -name "*.f.js" -not -path "./functions/lib/devUtils/*" -print | xargs -I{} basename {} ".f.js")

# Create an array to count functions reliably
FN_ARRAY=( $FUNCTION_NAMES )
NUM_FUNCTIONS=${#FN_ARRAY[@]}

echo "Found $NUM_FUNCTIONS functions to deploy"

if (( NUM_FUNCTIONS == 0 )); then
  echo "No functions found. Exiting."
  exit 0
fi


_deploy_batch() {
  local START_INDEX=$1
  local END_INDEX=$2
  DEPLOY_COMMAND=$(echo "$FUNCTION_NAMES" | sed -n "$START_INDEX,$END_INDEX p" | xargs | sed "s/ /,functions:/g;s%^%\.\/functions\/node_modules\/firebase-tools\/lib\/bin\/firebase.js deploy --force --only functions:%")
  echo "Executing command: $DEPLOY_COMMAND" | fold -w 120
  $DEPLOY_COMMAND
}

# Sequentially deploy all batches from a single node
start=1
while (( start <= NUM_FUNCTIONS )); do
  end=$(( start + BATCH_SIZE - 1 ))

  has_more_functions=true
  if (( end > NUM_FUNCTIONS )); then
    end=$NUM_FUNCTIONS;
    has_more_functions=false
  fi

  echo "Deploying functions $start through $end"
  _deploy_batch "$start" "$end"
  if $has_more_functions; then
    echo "Sleeping ${SLEEP_BETWEEN_BATCHES}s before next batch..."
    sleep "$SLEEP_BETWEEN_BATCHES"
  fi

  start=$(( end + 1 ))
done

echo "All batches deployed"
```

**Comment by davewd** [comment]
Well this is working better - thanks all for the help.  

Anyone have any good tricks on reducing fixed cost of load time ?  ie for a single function my code is taking 1m 47s  - whats the fastest with any form of caching e.t.c. ?  N.B. I asm deploying python based functions inside github actions incase that changes any approaches.

Max concurrent functions appears to be 20 for posterity.

**Comment by Zloka** [comment]
@inlined while there are workarounds mentioned in this issue, I opened PR #9252, which aims to catch 429 errors and retry with exponential backoff on the CLI level which, if merged and released, would allow us to drop the manual workarounds.

Pardon the ping, but this issue seems to affect many people and has stuck around for some time, and the PR hasn't been reviewed in a few weeks, so I wanted to make sure you at Firebase are aware of its existence 🙂
